File: d01/x2.py
#!/usr/bin/env python

numbers = {
    "zero": "0", "one": "1", "two": "2", "three": "3", "four": "4",
    "five": "5", "six": "6", "seven": "7", "eight": "8", "nine": "9",
    "0": "0", "1": "1", "2": "2", "3": "3", "4": "4",
    "5": "5", "6": "6", "7": "7", "8": "8", "9": "9",
}

def extract_numbers(line):
    firsti, lasti = len(line), -1
    first, last = None, None
    for k in numbers.keys():
        i = line.find(k)
        r = line.rfind(k)
        if i != -1 and i < firsti:
            firsti = i
            first = k
        if r != -1 and lasti < r:
            lasti = r
            last = k
    if first is None or last is None:
        raise Exception(f"error with: '{line}'; {first}, {last}")
    ret = [numbers[first], numbers[last]]
    print(f"{line.rstrip()} :: {ret}")
    return ret

with open("input.txt", "r") as f:
    lines = f.readlines()

numbers = [extract_numbers(line) for line in lines]
numbers = [int(''.join(number[:1] + number[-1:])) for number in numbers]
result = sum(numbers)

print(result)

File: d01/x2.rs
use std::collections::HashMap;
use std::fs;

fn extract_first_last(line: &str) -> (usize, usize) {
    let numerics = (0..=9).map(|x| x.to_string()).enumerate();
    let numerals = ["zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine"];
    let numerals = numerals.iter().map(|x| x.to_string()).enumerate();
    let numbers = numerals.chain(numerics).map(|(i, v)| (v, i)).collect::<HashMap<String, usize>>();

    let keys = numbers.keys().filter(|k| line.contains(&**k));

    let first = keys.clone().min_by_key(|k| line.find(&**k).unwrap());
    let last = keys.max_by_key(|k| line.rfind(&**k).unwrap());

    (*numbers.get(first.unwrap()).unwrap(), *numbers.get(last.unwrap()).unwrap())
}

fn main() {
    let contents = fs::read_to_string("input.txt").expect("read");
    let result: usize = contents
        .split('\n')
        .filter(|line| !line.is_empty())
        .inspect(|x| println!("about to filter: {x}"))
        .map(|line| extract_first_last(line))
        .inspect(|x| println!("about to filter: {x:?}"))
        .map(|(f, l)| f * 10 + l)
        .sum();
    println!("result: {result}");
}
